dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModelManager.java

```diff
@@ -0,0 +1,669 @@
+        }
+
+        Set<RelationGraph> relGraphs = new HashSet<>();
+        DependencyGraph<Relation> dependencyGraph = DependencyGraph.from(relsToExtract);
+
+        for (Set<DependencyGraph<Relation>.Node> component : dependencyGraph.getSCCs()) {
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = new RecursiveGraph();
+                    recGraph.setName(r.getNameOrTerm() + "_rec");
+                    relGraphs.add(recGraph);
+                    relGraphCache.put(r, recGraph);
+                }
+            }
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                RelationGraph rg = createGraph(r);
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = (RecursiveGraph) relGraphCache.get(r);
+                    recGraph.setConcreteGraph(rg);
+                } else {
+                    relGraphs.add(rg);
+                }
+            }
+        }
+
+        Set<CAATPredicate> predicates = new HashSet<>(relGraphs);
+        PredicateHierarchy hierarchy = new PredicateHierarchy(predicates);
+        hierarchy.initializeToDomain(domain);
+
+        // Populate graphs of base relations.
+        for (CAATPredicate basePred : hierarchy.getBasePredicates()) {
+            Relation r = relGraphCache.inverse().get((RelationGraph) basePred);
+            try {
+                r.getDefinition().accept(graphPopulator);
+            } catch (UnsupportedOperationException e) {
+                // Populate graph of relations unsupported by the visitor using default relation analysis.
+                graphPopulator.populateDynamicDefaultGraph(r);
+            }
+        }
+
+        // Do the computation.
+        hierarchy.populate();
+
+        for (CAATPredicate pred : hierarchy.getPredicateList()) {
+            Relation r = relGraphCache.inverse().get((RelationGraph) pred);
+            if (relModelCache.containsKey(r)) {
+                RelationModel rm = relModelCache.get(r);
+                ((RelationGraph) pred).edgeStream()
+                                      .forEach(e -> rm.addEdgeModel(getOrCreateEdgeModel(e)));
+           }
+        }
+
+        for (Relation r : relsToExtract) {
+            executionModel.addRelation(r, relModelCache.get(r));
+        }
+    }
+
+    private void createModel(Relation r) {
+        if (relModelCache.containsKey(r)) { return; }
+        relModelCache.put(r, new RelationModel(r));
+    }
+
+    private RelationGraph createGraph(Relation r) {
+        RelationGraph rg;
+        try {
+            rg = r.getDefinition().accept(graphBuilder);
+        } catch (UnsupportedOperationException e) {
+            // Generate a SimpleGraph for base relations that can be populated manually.
+            rg = new SimpleGraph();
+        }
```

Comment:
This looks sketchy. Can't you just check if the definition has dependencies? If it has not it is base and otherwise it is derived. Exception-based reasoning should be avoided if possible.